src/runtime/CollectionWrappers/ListWrapper.cs

```diff
@@ -0,0 +1,71 @@
+using System;
+using System.Collections.Generic;
+
+namespace Python.Runtime.CollectionWrappers
+{
+    internal class ListWrapper<T> : SequenceWrapper<T>, IList<T>
+    {
+        public ListWrapper(PyObject pyObj) : base(pyObj)
+        {
+
+        }
+
+        public T this[int index]
+        {
+            get
+            {
+                var item = Runtime.PyList_GetItem(pyObject.Reference, index);
+                var pyItem = new PyObject(item);
+
+                if (!Converter.ToManaged(pyItem.Handle, typeof(T), out object obj, true))
+                    Runtime.CheckExceptionOccurred();
+
+                return (T)obj;
+            }
+            set
+            {
+                IntPtr pyItem = Converter.ToPython(value, typeof(T));
+                if (pyItem == IntPtr.Zero)
+                {
+                    throw new InvalidCastException(
+                        "cannot cast " + value.ToString() + "to type: " + typeof(T).ToString(),
+                        new PythonException());
+                }
+
+                var result = Runtime.PyList_SetItem(pyObject.Handle, index, pyItem);
+                if (result == -1)
+                    Runtime.CheckExceptionOccurred();
+            }
+        }
+
+        public int IndexOf(T item)
+        {
+            return indexOf(item);
+        }
+
+        public void Insert(int index, T item)
+        {
+            if (IsReadOnly)
+                throw new InvalidOperationException("Collection is read-only");
+
+            IntPtr pyItem = Converter.ToPython(item, typeof(T));
+            if (pyItem == IntPtr.Zero)
+                throw new PythonException();
+
+            var result = Runtime.PyList_Insert(pyObject.Reference, index, pyItem);
+            Runtime.XDecref(pyItem);
```

Comment:
Doing a decref with error set might have unintended consequences.

Reply:
@lostmsu with value.ToPython() I don't have to manually mess with the ref count